File: disease/python/disease.py
# This script defines an abstraction for a player in the game, and several 
# helper functions.

import math, network, networkx, operator, os, pickle, random

class Player(network.Vertex):
    """
    Abstraction for an individual in the network. Each individual has an id, 
    a infection (infection probability), c (disease-specific death 
    probability), and state (susceptible, infected, recovered, vaccinated) 
    value, and also maintains a history of these values.
    """ 

    # Any player is in one of these states.
    SUSCEPTIBLE = 0
    INFECTED    = 1
    RECOVERED   = 2
    VACCINATED  = 3

    def __init__(self, id):
        """
        Construct a Player given the id. 
        """

        self.id = id
        self.infection = []
        self.c = []
        self.state = []
        self.set_infection(0.0)
        self.set_c(0.0)
        self.set_state(Player.SUSCEPTIBLE)

    def set_infection(self, v):
        """
        Set the transmission rate for this individual. 
        """

        self.infection_current = v
        
    def get_infection(self):
        """
        Return the transmission rate for this individual.
        """

        return self.infection_current

    def get_infection_list(self):
        """
        Return the transmission rate history of this individual.
        """

        return self.infection

    def set_c(self, v):
        """
        Set the disease-induced death rate to the specified value.
        """
        
        self.c_current = v

    def get_c(self):
        """
        Return the disease-induced death rate for this individual.
        """

        return self.c_current
    
    def get_c_list(self):
        """
        Return the disease-induced death rate history of this individual.
        """

        return self.c

    def set_state(self, v):
        """
        Set the state of this individual.
        """

        self.state_current = v

    def get_state(self):
        """
        Return the state of this individual.
        """

        return self.state_current

    def get_state_list(self):
        """
        Return the state history of this individual.
        """

        return self.state

    def save(self):
        """
        Record the transmission rate, disease-induced death rate, and the state 
        of this individual in the respective histories.
        """

        self.infection.append(self.infection_current)
        self.c.append(self.c_current)
        self.state.append(self.state_current)

    def __str__(self):
        """
        Return a string representation for this individual.
        """

        return "Player (id = %d, infection = %f, c = %f, state = %d)" \
            %(self.id, self.infection_current, self.c_current, 
              self.state_current)

def state_stats(infile):
    """
    Lists the number of individuals in different states in the results file 
    specified by infile.
    """

    fh = open(infile, "r")
    params = pickle.load(fh)
    population = pickle.load(fh)
    fh.close()
    s = 0
    i = 0
    r = 0
    v = 0
    for p in population:
        if p.get_state() == Player.SUSCEPTIBLE:
            s += 1
        elif p.get_state() == Player.INFECTED:
            i += 1
        elif p.get_state() == Player.RECOVERED:
            r += 1
        elif p.get_state() == Player.VACCINATED:
            v += 1
    print "N = %d, S = %d, I = %d, R = %d, V = %d" %(len(population), 
                                                     s, i, r, v)

def degree_vaccination(infile, outfile, k):
    """
    Change the state (to VACCINATED) of the top k (%) highly-connected 
    individuals in population contained in infile, and save the altered 
    population into outfile. 
    """

    fh = open(infile, "r")
    params = pickle.load(fh)
    population = pickle.load(fh)
    fh.close()
    net = network.build_network(population, params["network_topology"], 
                                params["network_params"])
    L = networkx.degree_centrality(net.get_structure()).items()
    L = sorted(L, key = operator.itemgetter(1), reverse = True)
    n = int(k / 100.0 * len(population))
    for i in range(0, n):
        person = population[L[i][0]]
        person.set_state(Player.VACCINATED)
        sl = person.get_state_list()
        sl[len(sl) - 1] = Player.VACCINATED
        person.set_infection(0.0)
        il = person.get_infection_list()
        il[len(il) - 1] = 0.0
        person.set_c(0.0)
        cl = person.get_c_list()
        cl[len(cl) - 1] = 0.0
    fh = open(outfile, "w")
    pickle.dump(params, fh)
    pickle.dump(population, fh)
    fh.close()

def referral_vaccination(infile, outfile, k):
    """
    Change the state (to VACCINATED) of k (%) individuals in population 
    contained in infile, and save the altered population into outfile. Each of 
    these individuals is a randomly picked neighbor of a randomly picked 
    person. 
    """

    fh = open(infile, "r")
    params = pickle.load(fh)
    population = pickle.load(fh)
    fh.close() 
    net = network.build_network(population, params["network_topology"], 
                                params["network_params"])
    m = int(k / 100.0 * len(population))
    i = 0
    while i < m:
        person = net.get_random_vertex()
        person = net.get_random_neighbor(person)
        if person == None or person.get_state() == Player.VACCINATED: 
            continue
        person.set_state(Player.VACCINATED)
        sl = person.get_state_list()
        sl[len(sl) - 1] = Player.VACCINATED
        person.set_infection(0.0)
        il = person.get_infection_list()
        il[len(il) - 1] = 0.0
        person.set_c(0.0)
        cl = person.get_c_list()
        cl[len(cl) - 1] = 0.0
        i += 1
    fh = open(outfile, "w")
    pickle.dump(params, fh)
    pickle.dump(population, fh)
    fh.close()

def random_vaccination(infile, outfile, k):
    """
    Change the state (to VACCINATED) of k (%) individuals in population 
    contained in infile, and save the altered population into outfile. Each of 
    these individuals is randomly picked. 
    """

    fh = open(infile, "r")
    params = pickle.load(fh)
    population = pickle.load(fh)
    fh.close()
    net = network.build_network(population, params["network_topology"], 
                                params["network_params"])
    m = int(k / 100.0 * len(population))
    i = 0
    while i < m:
        person = net.get_random_vertex()
        if person == None or person.get_state() == Player.VACCINATED:
            continue
        person.set_state(Player.VACCINATED)
        sl = person.get_state_list()
        sl[len(sl) - 1] = Player.VACCINATED
        person.set_infection(0.0)
        il = person.get_infection_list()
        il[len(il) - 1] = 0.0
        person.set_c(0.0)
        cl = person.get_c_list()
        cl[len(cl) - 1] = 0.0
        i += 1
    fh = open(outfile, "w")
    pickle.dump(params, fh)
    pickle.dump(population, fh)
    fh.close()

def fitness(infection, b, a):
    """
    Return the virulence given the infection probability, and constants 
    b and a.
    """
    
    return math.pow(infection / b, 1.0 / a)

def bins(items, low, high, n_bins):
    """
    Bin the elements in the items list into n_bins and return the list of 
    bins. The low and high values specifiy the minimum and maximum values for 
    the elements.
    """

    l = []
    delta = (high - low) / n_bins
    a = low
    b = a + delta
    for i in range(0, n_bins):
        n = len(filter(lambda x: x >= a and x < b, items))
        l.append(n)
        a = b
        b += delta
    return l

File: disease/java/Disease.java
// Disease.java

// Class that implements the agent-based virus dynamics model.
public class Disease
{
    // Every individual in the population is in one of these three states.
    private static int SUSCEPTIBLE = 0;
    private static int INFECTED = 1;
    private static int RECOVERED = 2;

    // Returns the virulence corresponding to the specified infectivity. 
    // The formula used is c = pow(infection / b, 1.0 / a).
    private static double fitness(double b, double a, double infection)
    {
	return Math.pow(infection / b, 1.0 / a);
    }
    
    // Returns a string representation of the items in the array a.
    private static String toString(int[] a)
    {
	String s = "";
	int i;
	for (i = 0; i < a.length - 1; i++) {
	    if (a[i] != 0) {
		s += i + ":" + a[i] + " "; 
	    }
	}
	s += (a[i] != 0) ? i + ":" + a[i] : "";
	return s;
    }

    // Entry point.
    public static void main(String[] args)
    {
	if (args.length != 1) {
	    System.out.println("Usage: java Disease <params file>");
	    System.exit(1);
	}
	Params params = Params.loadParams(args[0]);

	// States, infectivity values, and virulence values of individuals 
	// in the population.
	int[] states = new int[params.population];
	double[] infections = new double[params.population]; 
	double[] cs = new double[params.population];    

	// The underlying network structure.
	Network net;
	if (params.network_topology.equals("Complete")) {
	    net = new CompleteNetwork(params.population);
	}
	else {
	    net = new GraphMLNetwork(params.graphml_file);
	}

	// Infect an individual at random. This individual is the random 
	// neighbor of a randomly picked individual.
	int i, j;
	do {
	    i = net.getRandomVertex();
	}
	while ((j = net.getRandomNeighbor(i)) == -1);
	states[j] = INFECTED;
	infections[j] = params.infection;
	cs[j] = fitness(params.b, params.a, params.infection);

	System.out.printf("%d %d %d\n", params.population, params.generations, 
			  params.report_freq);

	// The dynamics.
	for (int t = 0; t <= params.generations; t++) {

	    // Gather and print some stats at t = report_freq. These include: 
	    // the number of susceptible and infected individuals, 
	    // the mean virulence level, and the distribution of virulence.
	    if (t % params.report_freq == 0) {
		int nS = 0;
		int nI = 0;
		int[] dist = new int[101];
		for (int p = 0; p < params.population; p++) {
		    if (states[p] == SUSCEPTIBLE) {
			nS += 1;
		    }
		    else if (states[p] == INFECTED) {
			nI += 1;
			dist[(int) Math.floor(cs[p] * 100)]++;
		    }
		}
		System.out.printf("%d %d %5.4f %s\n", nS, nI, 
				  Stats.sum(cs) / nI, 
				  toString(dist));
	    }

	    for (int p = 0; p < params.population; p++) {
		i = net.getRandomVertex();

		// i is susceptible.
		if (states[i] == SUSCEPTIBLE) {
		    if (Random.uniform() < params.death) {
			// i dies of natural cuases and is reborn as a 
			// susceptible individual.
			continue;
		    }
		    
		    // Find neighbor that could infect i. 
		    double probNot = 1.0;
		    double totalInfection = 0.0;
		    int[] neighbors = net.getNeighbors(i);
		    if (neighbors == null) {
			continue;
		    }
		    for (int k: neighbors) {
			if (states[k] == INFECTED) {
			    probNot *= (1.0 - infections[k]);
			    totalInfection += infections[k];
			}
		    }

		    if (Random.uniform() < 1.0 - probNot) {
			// i is infected by a neighbor, and takes on that 
			// individual's (possibly mutated) infectivity value.
			states[i] = INFECTED;
			double newInfection = 0.0;
			double r = Random.uniform() * totalInfection;
			double sum = 0.0;
			for (int k: neighbors) {
			    newInfection = infections[k];
			    sum += newInfection;
			    if (r < sum) {
				break;
			    }
			}
			if (Random.uniform() < params.mutation) {
			    newInfection = Math.max(0.0, 
					       Math.min(Random.
							gaussian(newInfection,
								 params.stddev),
							1.0));
			}
			infections[i] = newInfection;
			cs[i] = fitness(params.b, params.a, newInfection);
		    }
		}
		// i is infected.
		else if (states[i] == INFECTED) {
		    if (Random.uniform() < 
			1.0 - (1.0 - params.death) * (1.0 - cs[i])) {
			// i dies either of natural causes or due to the 
			// virus, and is reborn as a susceptible individual.
			states[i] = SUSCEPTIBLE;
			infections[i] = 0.0;
			cs[i] = 0.0;
		    }
		    else if (Random.uniform() < params.recovery) {
			// i recovers from the virus.
			states[i] = RECOVERED;
			infections[i] = 0.0;
			cs[i] = 0.0;
		    }
		}
		// i is recovered.
		else if (states[i] == RECOVERED) {
		    if (Random.uniform() < params.death) {
			// i dies of natural causes and is reborn as a 
			// susceptible individual.
			states[i] = SUSCEPTIBLE;
			infections[i] = 0.0;
			cs[i] = 0.0;
		    }
		}
	    }
	}
    }
}